tachyon/crypto/commitments/kzg/shplonk.h

```diff
@@ -178,22 +187,173 @@ class SHPlonk : public UnivariatePolynomialCommitmentScheme<
     return writer->WriteToProof(q);
   }
 
+  template <typename ContainerTy>
+  [[nodiscard]] bool DoVerifyOpeningProof(
+      const ContainerTy& poly_openings,
+      TranscriptReader<Commitment>* reader) const {
+    using G1JacobianPointTy = typename G1PointTy::JacobianPointTy;
+
+    Field y = reader->SqueezeChallenge();
+    Field v = reader->SqueezeChallenge();
+
+    Commitment h;
+    if (!reader->ReadFromProof(&h)) return false;
+
+    Field u = reader->SqueezeChallenge();
+
+    Commitment q;
+    if (!reader->ReadFromProof(&q)) return false;
+
+    PolynomialOpeningGrouper<Poly, Commitment> grouper;
+    grouper.GroupByPolyAndPoints(poly_openings);
+
+    // Group |poly_openings| to |grouped_poly_openings_vec|.
+    // {[C₀, C₁, C₂], [x₀, x₁, x₂]}
+    // {[C₃], [x₂, x₃]}
+    // {[C₄], [x₄]}
+    const std::vector<GroupedPolynomialOpenings<Poly, Commitment>>&
+        grouped_poly_openings_vec = grouper.grouped_poly_openings_vec();
+    const absl::btree_set<PointDeepRef>& super_point_set =
+        grouper.super_point_set();
+
+    Field first_z_diff_inverse = Field::Zero();
+    Field first_z = Field::Zero();
+
+    std::vector<G1JacobianPointTy> l_commitments;
+    l_commitments.reserve(grouped_poly_openings_vec.size());
+    size_t i = 0;
+    for (const auto& [poly_openings_vec, points_ref] :
```

Comment:
How about
```suggestion
    for (const auto& [poly_openings_vec, points] :
```
or rename member name in `GroupedPolynomialOpenings`?

Reply:
Agreed. I will rename the member name to `point_refs`